pkg/apis/machine/v1beta1/machineset_webhook.go

```diff
@@ -0,0 +1,147 @@
+	decoder           *admission.Decoder
+}
+
+// NewMachineSetValidator returns a new machineSetValidatorHandler.
+func NewMachineSetValidator() (*machineSetValidatorHandler, error) {
+	infra, err := getInfra()
+	if err != nil {
+		return nil, err
+	}
+
+	return createMachineSetValidator(infra.Status.PlatformStatus.Type, infra.Status.InfrastructureName), nil
+}
+
+func createMachineSetValidator(platform osconfigv1.PlatformType, clusterID string) *machineSetValidatorHandler {
+	return &machineSetValidatorHandler{
+		clusterID:         clusterID,
+		webhookOperations: getMachineValidatorOperation(platform),
+	}
+}
+
+// NewMachineSetDefaulter returns a new machineSetDefaulterHandler.
+func NewMachineSetDefaulter() (*machineSetDefaulterHandler, error) {
+	infra, err := getInfra()
+	if err != nil {
+		return nil, err
+	}
+
+	return createMachineSetDefaulter(infra.Status.PlatformStatus.Type, infra.Status.InfrastructureName), nil
+}
+
+func createMachineSetDefaulter(platform osconfigv1.PlatformType, clusterID string) *machineSetDefaulterHandler {
+	return &machineSetDefaulterHandler{
+		clusterID:         clusterID,
+		webhookOperations: getMachineDefaulterOperation(platform),
+	}
+}
+
+// InjectDecoder injects the decoder.
+func (v *machineSetValidatorHandler) InjectDecoder(d *admission.Decoder) error {
+	v.decoder = d
+	return nil
+}
+
+// InjectDecoder injects the decoder.
+func (v *machineSetDefaulterHandler) InjectDecoder(d *admission.Decoder) error {
+	v.decoder = d
+	return nil
+}
+
+// Handle handles HTTP requests for admission webhook servers.
+func (h *machineSetValidatorHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	ms := &MachineSet{}
+
+	if err := h.decoder.Decode(req, ms); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	klog.V(3).Infof("Validate webhook called for MachineSet: %s", ms.GetName())
+
+	if ok, err := h.validateMachineSet(ms); !ok {
+		return admission.Denied(err.Error())
+	}
+
+	return admission.Allowed("MachineSet valid")
+}
+
+// Handle handles HTTP requests for admission webhook servers.
+func (h *machineSetDefaulterHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	ms := &MachineSet{}
+
+	if err := h.decoder.Decode(req, ms); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	klog.V(3).Infof("Mutate webhook called for MachineSet: %s", ms.GetName())
+
+	if ok, err := h.defaultMachineSet(ms); !ok {
+		return admission.Denied(err.Error())
+	}
+
+	marshaledMachineSet, err := json.Marshal(ms)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledMachineSet)
+}
+
+func (h *machineSetValidatorHandler) validateMachineSet(ms *MachineSet) (bool, utilerrors.Aggregate) {
+	var errs []error
+
+	// Create a Machine from the MachineSet and validate the Machine template
+	m := &Machine{Spec: ms.Spec.Template.Spec}
```

Comment:
I would add a validator that `selector.matchLabels` is equal to machine labels.

Reply:
We had planned for this PR to just be providing validation for the ProviderSpec. I'm intending to follow up next sprint with a validation for the other fields within MachineSetSpec, this being one of those validations